src/widget_impl/lTerm_scroll_impl.ml

```diff
@@ -0,0 +1,187 @@
+    method update_resources =
+      let rc = self#resource_class and resources = self#resources in
+      focused_style <- LTerm_resources.get_style (rc ^ ".focused") resources;
+      unfocused_style <- LTerm_resources.get_style (rc ^ ".unfocused") resources
+
+    val mutable range = 0
+    val mutable offset = 0
+    method range = range
+    method set_range r = range <- max 0 r
+    method offset = offset
+    method set_offset o = offset <- max 0 (min (range-1) o); self#queue_draw
+    method scroll_bar_size = default_scroll_bar_size
+
+    method virtual private mouse_offset : LTerm_mouse.t -> rect -> int
+    method virtual private key_scroll_incr : LTerm_key.code
+    method virtual private key_scroll_decr : LTerm_key.code
+    method virtual private scroll_window_size : int
+
+    method private scroll_bar_steps = 
+      self#scroll_window_size - self#scroll_bar_size + 1
+
+    val mutable scroll_bar_offset = 0
+    method private set_scroll_bar_offset o = 
+      scroll_bar_offset <- max 0 (min (self#scroll_bar_steps-1) o)
+
+    method private window_of_scroll offset = 
+      self#set_scroll_bar_offset offset;
+      map_range (self#scroll_bar_steps-1) (range-1) scroll_bar_offset
+
+    method private scroll_of_window offset = 
+      let offset = map_range (range-1) (self#scroll_bar_steps-1) offset in
+      self#set_scroll_bar_offset offset;
+      scroll_bar_offset
+
+    method incr = 
+      if range >= self#scroll_bar_steps then
+        self#set_offset @@ self#window_of_scroll (scroll_bar_offset+1) 
+      else
+        self#set_offset (self#offset+1);
+
+    method decr = 
+      if range >= self#scroll_bar_steps then
+        self#set_offset @@ self#window_of_scroll (scroll_bar_offset-1) 
+      else
+        self#set_offset (self#offset-1);
+
+  initializer self#on_event @@ fun ev ->
+    let open LTerm_mouse in
+    let open LTerm_key in
+
+    let alloc = self#allocation in
+
+    match ev with
+    | LTerm_event.Mouse m when m.button=Button1 && in_rect alloc (coord m) ->
+      let scroll = self#mouse_offset m alloc in
+      self#set_offset @@ self#window_of_scroll scroll;
+      true
+
+    | LTerm_event.Key { control = false; meta = false; shift = true; code } 
+      when code=self#key_scroll_decr ->
+      self#decr;
+      true
+
+    | LTerm_event.Key { control = false; meta = false; shift = true; code } 
+      when code=self#key_scroll_incr ->
+      self#incr;
+      true
+
+    | _ -> false
+
+  end
+
+class vscrollbar ?size_request ?(default_scroll_bar_size=5) () = object(self)
+  inherit scrollbar ~default_scroll_bar_size
+
+  method size_request = 
+    match size_request with None -> { rows=0; cols=2 } | Some(x) -> x
+
+  method private mouse_offset m alloc = m.LTerm_mouse.row - alloc.row1 
+  method private key_scroll_incr = LTerm_key.Down
+  method private key_scroll_decr = LTerm_key.Up
+  method private scroll_window_size = self#allocation.row2 - self#allocation.row1
```

Comment:
It's not a big deal, but it's generally better to save the result of the method call:

```ocaml
   let rect = self#allocation in
   ...
```

otherwise the compiler will issue the method call twice, as it has no way of knowing they'll produce the same result